### ISSUE TITLE
Inconsistent results with table service: source location is moved in certain queries

### ISSUE THREAD

**Issue by erikvullings** [Project-OSRM/osrm-backend#5834]
In order to calculate the driving distances (isochrones) from a certain point, I use the table service to query multiple points from a single source (origin) to all destinations, which are in several circles around the origin. This algorithm is implemented in the [howfar](https://www.npmjs.com/package/howfar) service, and basically works fine. However, under certain conditions, the results are unexpected: the origin location is at a small island in the north of The Netherlands (Schiermonnikoog), and the maximum reach is 25km, crossing the water to the mainland. In that case, the source location is moved from the island to the mainland, as can be seen from this query!

```bash
https://router.project-osrm.org//table/v1/driving/6.16808,53.479854;6.167965,53.479973;6.16788,53.479923;6.167849,53.479854;6.16788,53.479785;6.167965,53.479735;6.16808,53.479717;6.168195,53.479735;6.16828,53.479785;6.168311,53.479854;6.16828,53.479923;6.168195,53.479973;6.16808,53.479991;6.167602,53.480916;6.166774,53.480631;6.166296,53.480139;6.166296,53.479569;6.166774,53.479077;6.167602,53.478792;6.168558,53.478792;6.169386,53.479077;6.169864,53.479569;6.169864,53.480139;6.169386,53.480631;6.168558,53.480916;6.164963,53.483067;6.162681,53.481709;6.161846,53.479854;6.162681,53.477999;6.164963,53.476641;6.16808,53.476144;6.171197,53.476641;6.173479,53.477999;6.174314,53.479854;6.173479,53.481709;6.171197,53.483067;6.16808,53.483564;6.164254,53.488349;6.157629,53.486072;6.153805,53.482129;6.153806,53.477577;6.157632,53.473635;6.164256,53.471359;6.171904,53.471359;6.178528,53.473635;6.182354,53.477577;6.182355,53.482129;6.178531,53.486072;6.171905,53.488349;6.153643,53.494729;6.143079,53.48844;6.139217,53.479851;6.143089,53.471263;6.153653,53.464978;6.16808,53.462677;6.182507,53.464978;6.193071,53.471263;6.196943,53.479851;6.193081,53.48844;6.182517,53.494729;6.16808,53.497031;6.155158,53.508523;6.132793,53.500836;6.119895,53.487526;6.119912,53.472163;6.132827,53.458862;6.155176,53.451184;6.180976,53.451183;6.203328,53.45886;6.216247,53.472162;6.216265,53.487527;6.203362,53.500838;6.180993,53.508524.json?sources=0&generate_hints=false
```
The result shows that the origin/source has moved from `[6.16808,53.479854;6]` to `[6.17875,53.414216]`:
```json
{ "sources":[{"name":"Bantswei","location":[6.17875,53.414216]}] }
```

When I make the reach (maximum circle radius) significantly smaller, everything is fine: 
```bash
https://router.project-osrm.org/table/v1/driving/6.16808,53.479854;6.167205,53.480755;6.166565,53.480374;6.166331,53.479854;6.166565,53.479334;6.167205,53.478953;6.16808,53.478813;6.168955,53.478953;6.169595,53.479334;6.169829,53.479854;6.169595,53.480374;6.168955,53.480755;6.16808,53.480895;6.164457,53.487897;6.158184,53.485742;6.154564,53.482008;6.154565,53.477698;6.158187,53.473966;6.164459,53.471811;6.171701,53.471811;6.177973,53.473965;6.181595,53.477698;6.181596,53.482009;6.177975,53.485742;6.171702,53.487897;6.144454,53.50419;6.127169,53.493899;6.120855,53.479845;6.127196,53.465795;6.144481,53.455513;6.16808,53.45175;6.191679,53.455513;6.208964,53.465795;6.215305,53.479845;6.208991,53.493899;6.191706,53.50419;6.16808,53.507958;6.139043,53.544194;6.088828,53.526927;6.059909,53.497045;6.059996,53.462569;6.089001,53.432731;6.13913,53.415508;6.196988,53.415501;6.247136,53.432717;6.276161,53.462562;6.276248,53.497052;6.247309,53.52694;6.197074,53.5442;6.058472,53.592483;5.978447,53.544759;5.949447,53.479655;5.979028,53.41465;6.059053,53.367126;6.16808,53.349744;6.277107,53.367126;6.357132,53.41465;6.386713,53.479655;6.357713,53.544759;6.277688,53.592483;6.16808,53.609964;6.06956,53.696946;5.89982,53.638467;5.802648,53.537469;5.803631,53.421167;5.901787,53.320666;6.070548,53.262685;6.265132,53.262608;6.434113,53.320512;6.532493,53.421088;6.533478,53.537545;6.436082,53.638621;6.266113,53.697023.json?sources=0&generate_hints=false
```
As you can also see from the result, where the source location has remained the same:
```json
{ "sources": [{"name":"","location":[6.16808,53.479854]}] }
```

What is happening here? Why has the source location moved in the first case, and not in the second?

As a further note: I find it curious that, when querying the table distance/duration across the water, a result is returned, whereas the route service does not generate any route. I would have preferred if the table result would have indicated that the location is not reachable (perhaps using a duration of -1).

**Comment by danpat** [comment]
OSRM utilizes [strongly connected components](https://en.wikipedia.org/wiki/Strongly_connected_component) to help guide the snapping location.  Basically, it attempts to avoid routing failures by always snapping both coordinates to a part of the graph that is known to be connected.

There's an old blog post by Denis here that describes the behaviour: https://blog.mapbox.com/robust-navigation-with-smart-nearest-neighbor-search-dbc1f6218be8

Basically, your island is not connected to the mainland, so it's not "strongly connected" - because *other* points are on the mainland and part of a large network, your origin location is being moved to ensure you get routing results.

There is currently no way to disable this behaviour.

**Comment by erikvullings** [comment]
@danpat thanks for the explanation. I get what's happening, but still find it odd that you apply this concept also to table queries, and furthermore, accept a jump of 7km of the source location. There should be limits to this jump - it's too aggressive. Currently, i have to iteratively check for jumps, and if the jump is too big (e.g. >1km), I try to compute the isochrone again using a smaller radius. Not ideal, since it means that I miss nearby locations if the island is not round.
